### ISSUE TITLE
Create type safe names/identifiers/literals in DAML engine 

### ISSUE THREAD

**Issue by remyhaemmerle-da** [digital-asset/daml#705]
We should improve type safety of daml-lf data. 

1. We should drop string  type aliases in `com.digitalasset.daml.lf.data.Ref` in favor of Boxed string whose contains are check at creation.

    This includes:
    * `PackageId` boxes a string that matches   ` [a-zA-Z0-9\-_ ]+`
    * Party  boxes a string that matches ` [a-zA-Z0-9\-_ ]+`
    * ChoiceName   boxes a string that matches  `[a-zA-Z_\$][a-zA-Z0-9_\$]*`
    * FieldName boxes a string that matches `[a-zA-Z_\$][a-zA-Z0-9_\$]*`
    * VariantConstructorName boxes a string that matches `[a-zA-Z_\$][a-zA-Z0-9_\$]*`
    * AbsoluteContractId boxes a string that matches ` [A-Za-z0-9._:-]+`

2. We should drop Decimal alias in favor of a boxed BigDecimal.

~~3. We should also encapsulate UTF8String (used inside Text value and key maps) to be sure that those are always processed (for sorting, exploding) in a UTF8 compatible way.~~

~~4. We should create a separate type for `ModuleName` and `EntityName`~~

**Comment by bitonic** [comment]
@remyhaemmerle-da for what regards strings, do you propose to actually have a different newtype for each of those? or a single newtype + tags? point 2 and 3 look good.

**Comment by S11001001** [comment]
How important is Java support? We can do a nice abstract-subtypes-of-String for your regex-constrained suggestions and maybe something similar for your other ideas depending on how you want the type relationships to go, if we enforce that in Scala only. If it is-a String, you don't have to call `underlyingString`.

**Comment by remyhaemmerle-da** [comment]
Inded, I propose to have a new type for each of the boxed strings.
I do not see how it can hurt. We actually do not need they have the same type + we can steel share the code if we want.

**Comment by remyhaemmerle-da** [comment]
@stephencompall-DA 
For UTF8 String, I was thinking to use AnyVal since there is a bijection between UTF16 strings  and UTF8 strings.

For the others, I am not  sure. It is really depending on how is important the Java support.

**Comment by S11001001** [comment]
This isn't a good use case for AnyVal regardless; if you want Java representation it won't work, if you want it to be erased AnyVal won't work and you need to use abstract types.

**Comment by S11001001** [comment]
To clarify: `AnyVal` is never a good choice for newtypes.

**Comment by remyhaemmerle-da** [comment]
@stephencompall-DA 
I will consult you before starting any subtype implementation.

**Comment by S11001001** [comment]
Please consult @S11001001 instead of @stephencompall-DA  😆

**Comment by remyhaemmerle-da** [comment]
@S11001001 
What it is your idead of abstract-subtypes-of-String for the regex-constrained suggestion ?

**Comment by bitonic** [comment]
@remyhaemmerle-da my fear with what regards the "many newtypes" plan is that they'll become so unbearable to use that people will just use them when they _must_ but use `String` everywhere. so I think it's worth thinking whether we can identify a smaller set of string types that we can reuse.

for what concerns Java: these library are used to implement the engine / ledger server, so Java support is not important.

**Comment by S11001001** [comment]
@remyhaemmerle-da As the idea is for each newtype in your first category to follow some regex, simply parameterizing the creation of newtypes by a Regex stands up well to factoring:

```scala
sealed abstract class MatchingStringModule {
  type T <: String
  def apply(s: String): Option[T]
  def verify[F[_]: Foldable](fs: F[String]): Either[String, F[T]]
}

object MatchingStringModule extends (Regex => MatchingStringModule) {
  def apply(r: Regex): MatchingStringModule = new MatchingStringModule {
    type T = String
    // rest of implementation proves that Strings match before making them Ts
  }
}

// elsewhere
val PackageId = MatchingStringModule(raw"[a-zA-Z0-9\-_ ]+".r)
type PackageId = PackageId.T
```

Where `verify` is [akin to narrow here](https://failex.blogspot.com/2017/04/and-glorious-subst-to-come.html#markdown-header-cheap-tagging-with-validation).

**Comment by remyhaemmerle-da** [comment]
Thanks.

**Comment by remyhaemmerle-da** [comment]
This issue should fix #398
